### PR TITLE
Add forge abstraction layer + adapt to GitHub

### PR DESCRIPTION
**Pull request by nphilipp**
```
commit c313cd60f67bebd3e652db82eb4718d18235b8a3
Author: Nils Philippsen <nils@redhat.com>
Date:   Wed Jul 31 12:20:43 2024 +0200

    Improve docstrings
    
    Use sphinx format to document parameters, etc., and fix some mistakes
    and wording.
    
    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit cc81bb5272c941d4ae1fb6b71cb562bdf7ea84b7
Author: Nils Philippsen <nils@redhat.com>
Date:   Thu Jul 18 16:21:44 2024 +0200

    Add forge abstraction layer
    
    Previously, one object was responsible for handling the communication
    with Pagure concerning all configured repositories.
    
    After the refactoring, working with forge APIs goes through the
    (abstract) Repository class, which allows creating concrete repository
    objects from their respective configuration snippets. The
    Repository.from_config() class method dispatches to the responsible
    child class, depending on the `type` configuration key.
    
    In the course, revamp the configuration format and add a configuration
    subsystem, which has these features over the old one:
    
    - multiple upstream instances
    - optional per-repo settings of enabled, label, blocked_label
    
    Additionally:
    
    - Introduce functions to read, validate and normalize configuration.
    - Use “status” consistently (instead of both “status” and “state”),
      except where dealing with the old configuration format.
    
    Related: #8
    
    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit 72422d463019f2b31592f50887a0e7fbad1f0b54
Author: Nils Philippsen <nils@redhat.com>
Date:   Wed Jul 31 12:43:44 2024 +0200

    Add --dry-run option
    
    Retrieve information from configured upstream repositories, but don’t
    attempt to change anything in JIRA.
    
    Related: #8
    
    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit 22fe0c9f50b85c847fc40b578d110f6afe2ff5f2
Author: Nils Philippsen <nils@redhat.com>
Date:   Thu Aug 1 17:27:24 2024 +0200

    Add querying GitHub repositories
    
    Fixes: #8
    
    Signed-off-by: Nils Philippsen <nils@redhat.com>
```

### CODE REVIEW

**Comment by nphilipp** [comment]
Want to add a config file format converter before merging.

File: config-legacy.example.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Example config for jira_sync
```
Comment (Zlopez) [inline config-legacy.example.toml:1]: The legacy config could be removed completely.

File: config.example.toml [hunk 2]
```diff
@@ -1,38 +1,38 @@
 # Example config for jira_sync
```
Comment (Zlopez) [inline config.example.toml:1]: This needs to be reflected in GitHub workflow as well.

File: jira_sync/config/legacy_model.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+"""Pydantic model describing the legacy configuration file."""
```
Comment (Zlopez) [inline jira_sync/config/legacy_model.py:1]: No need to keep legacy as well as we are the only one using that right now.

File: config-legacy.example.toml (hunk 1 above)
Comment (nphilipp) [reply config-legacy.example.toml]: 👍

File: jira_sync/config/legacy_model.py (hunk 3 above)
Comment (nphilipp) [reply jira_sync/config/legacy_model.py]: 👍

File: jira_sync/jira_wrapper.py [hunk 4]
```diff
@@ -17,57 +17,63 @@
...
-        url: str,
-        token: str,
+        url: str | None,
+        token: str | None,
         project: str,
         issue_type: str,
     ):
         """
-        Class constructor.
+        Object constructor.
```
Comment (Zlopez) [inline jira_sync/jira_wrapper.py:43]: I always thought that Object is instance of Class and constructor is creating the instance as you can't call it on existing object.

**Review by Zlopez — COMMENTED** [review]
Just one comment, but I would like to see github workflow that will run the tests on each PR. So we knew if anything is broken.

File: jira_sync/jira_wrapper.py (hunk 4 above)
Comment (nphilipp) [reply jira_sync/jira_wrapper.py:43]: Should have changed it to e.g. “Initialize the instance.” instead.

The term “constructor” is a bit out of place in Python – the `__new__()` class method creates an object (can even make it of a different class, whee) and the `__init__()` instance method initializes the created object.

Comment (Zlopez) [reply jira_sync/jira_wrapper.py:43]: I know that `constructor` is not really a correct term in Python, but I was learning OOP on Java.

**Pull request merged**